src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -86491,5 +86474,89 @@ public void M(TypedReference r)
+@"using System;
+
+class C
+{
+    static T F<T>(Func<T> f) => throw null!;
+    static void G(object? o)
+    {
+        F(() => o).ToString(); // warning: maybe null
+        if (o == null) return;
+        F(() => o).ToString(); // no warning
+    }
+}
+";
+            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (8,9): warning CS8602: Possible dereference of a null reference.
+                //         F(() => o).ToString(); // warning: maybe null
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F(() => o)").WithLocation(8, 9));
+        }
+
+        [Fact, WorkItem(29617, "https://github.com/dotnet/roslyn/issues/29617")]
+        public void CaptureVariablesWhereLambdaAppears_02()
+        {
+            var source =
+@"using System;
+
+class C
+{
+    static bool M(object? o) => throw null!;
+    static T F<T>(Func<T> f, bool ignored) => throw null!;
+    static void G(object? o)
+    {
+        F(() => o, M(1)).ToString(); // warning: maybe null
+        if (o == null) return;
+        F(() => o, M(o = null)).ToString(); // no warning
+    }
+}
+";
+            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (9,9): warning CS8602: Possible dereference of a null reference.
+                //         F(() => o, M(1)).ToString(); // warning: maybe null
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F(() => o, M(1))").WithLocation(9, 9));
+        }
+
+        [Fact, WorkItem(29617, "https://github.com/dotnet/roslyn/issues/29617")]
+        public void CaptureVariablesWhereLambdaAppears_03()
+        {
+            var source =
+@"using System;
+
+class C
+{
+    static bool M(object? o) => throw null!;
+    static Func<T> F<T>(Func<T> f, bool ignored = false) => throw null!;
+    static void G(object? o)
+    {
+        var fa1 = new[] { F(() => o), F(() => o, M(o = null)) };
+        fa1[0]().ToString(); // warning
+
+        if (o == null) return;
+        var fa2 = new[] { F(() => o), F(() => o, M(o = null)) };
+        fa2[0]().ToString();
+
+        if (o == null) return;
+        var fa3 = new[] { F(() => o, M(o = null)), F(() => o) };
+        fa3[0]().ToString(); // warning
+    }
+}
+";
+            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (10,9): warning CS8602: Possible dereference of a null reference.
+                //         fa1[0]().ToString(); // warning
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "fa1[0]()").WithLocation(10, 9),
+                // (18,9): warning CS8602: Possible dereference of a null reference.
+                //         fa3[0]().ToString(); // warning
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "fa3[0]()").WithLocation(18, 9));
+        }
```

Comment:
>} [](start = 8, length = 1)

Could you also test lambdas with side-effects on captured locals? My understanding is that we drop those side-effects (since we have no idea what order they'll be evaluated in).

For example: `F(() => x = null, () => x.ToString())`